### PR TITLE
Local Activity Cancellations

### PR DESCRIPTION
**Pull request by Sushisource**
Implement cancellation of local activities

### CODE REVIEW

**Review by cretz — APPROVED** [review]
Approving based on discussions on the previous incarnation at #241

File: src/test_help/history_builder.rs [hunk 1]
```diff
@@ -308,6 +308,24 @@ impl TestHistoryBuilder {
         self.add_local_activity_marker(seq, activity_id, None, Some(failure), None);
     }
 
+    pub(crate) fn add_local_activity_cancel_marker(&mut self, seq: u32, activity_id: &str) {
+        self.add_local_activity_marker(
+            seq,
+            activity_id,
+            None,
+            Some(Failure {
+                message: "cancelled bro".to_string(),
```
Comment (bergundy) [inline src/test_help/history_builder.rs:317]: 😆

**Review by bergundy — COMMENTED** [review]
I reviewed everything but `worker/activities/local_activities.rs`.

We should test all of the edge cases like activity completes before cancelled and cancelled before it completes.
I'd also make sure we have tests for permits being returned in all of these cases.

File: protos/local/workflow_commands.proto [hunk 2]
```diff
@@ -95,24 +93,31 @@ message ScheduleLocalActivity {
...
     /// Specify a retry policy for the local activity. By default local activities will be retried
     /// indefinitely.
     common.RetryPolicy retry_policy = 10;
     /// If the activity is retrying and backoff would exceed this value, lang will be told to
     /// schedule a timer and retry the activity after. Otherwise, backoff will happen internally in
-    /// core.
+    /// core. Defaults to 1 minute.
     google.protobuf.Duration local_retry_threshold = 11;
+    /// Defines how the workflow will wait (or not) for cancellation of the activity to be
+    /// confirmed. Lang should default this to `WAIT_CANCELLATION_COMPLETED`, even though proto
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply protos/local/workflow_commands.proto:118]: I'm not entirely sure, but I do think it makes sense for local activities. I would want to encourage people to cancel them cleanly.

File: src/machines/local_activity_state_machine.rs [hunk 3]
```diff
@@ -287,12 +357,41 @@ impl MarkerCommandCreated {
...
 #[derive(Default, Clone)]
 pub(super) struct Replaying {}
 impl Replaying {
     pub(super) fn on_schedule(self) -> LocalActivityMachineTransition<WaitingMarkerEvent> {
-        TransitionResult::ok([], WaitingMarkerEvent {})
+        TransitionResult::ok(
+            [],
+            WaitingMarkerEvent {
+                already_resolved: false,
+            },
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs:393]: It would over-complicate the transitions, having this bit of internal state works well for this purpose I think

File: src/machines/local_activity_state_machine.rs [hunk 4]
```diff
@@ -389,6 +515,40 @@ impl WaitingMarkerEvent {
...
+        )
+    }
+
+    fn on_no_wait_cancel(
+        self,
+        _: ActivityCancellationType,
+    ) -> LocalActivityMachineTransition<Abandoned> {
+        // There is nothing to be done here. We are waiting for a marker that will never come and
+        // we don't care about the result anyway.
+        TransitionResult::default()
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs:534]: Because this is on the replay path. There is nothing to request cancellation of, because nothing is executing.

File: src/machines/local_activity_state_machine.rs [hunk 5]
```diff
@@ -389,6 +515,40 @@ impl WaitingMarkerEvent {
...
+        TransitionResult::ok(
+            [LocalActivityCommand::RequestCancel],
+            WaitingMarkerEventCancelled {},
+        )
+    }
+
+    fn on_no_wait_cancel(
+        self,
+        _: ActivityCancellationType,
+    ) -> LocalActivityMachineTransition<Abandoned> {
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs:531]: Both of them "abandon" the activity in the sense that they no longer care about result information

File: src/machines/local_activity_state_machine.rs [hunk 6]
```diff
@@ -63,11 +77,19 @@ fsm! {
     // even though we already resolved the activity.
     WaitingMarkerEventPreResolved --(MarkerRecorded(CompleteLocalActivityData),
                                      shared on_marker_recorded) --> MarkerCommandRecorded;
+
+    // Ignore cancellation in final state
+    MarkerCommandRecorded --(Cancel, on_cancel_requested) --> MarkerCommandRecorded;
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs:82]: It will be discarded by the local activity manager since the activity is no longer tracked

File: src/machines/local_activity_state_machine.rs [hunk 7]
```diff
@@ -78,19 +100,41 @@ pub(super) struct ResolveDat {
...
-                Err(fail) => LocalActivityExecutionResult::Failed(ActFail {
-                    failure: Some(fail),
-                }),
+                Err(fail) => {
+                    if matches!(fail.failure_info, Some(FailureInfo::CanceledFailureInfo(_))) {
+                        LocalActivityExecutionResult::Cancelled {
+                            cancel: Cancellation {
+                                failure: Some(fail),
+                            },
+                            do_not_record_marker: true,
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs:130]: After recording markers for abandons, this is only true for cancels in the same WFT as invoking the activity

File: src/machines/local_activity_state_machine.rs [hunk 8]
```diff
@@ -198,47 +244,67 @@ impl LocalActivityMachine {
...
-    attrs: ScheduleLocalActivity,
+    attrs: ValidScheduleLA,
     replaying_when_invoked: bool,
     wf_time_when_started: Option<SystemTime>,
 }
 
+impl SharedState {
+    fn produce_no_wait_cancel_resolve_dat(&self) -> ResolveDat {
+        ResolveDat {
+            result: LocalActivityExecutionResult::empty_cancel(true),
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs]: We always do, now.

File: src/machines/local_activity_state_machine.rs [hunk 9]
```diff
@@ -53,6 +60,13 @@ fsm! {
     // to eventually see the marker
     WaitingMarkerEvent --(MarkerRecorded(CompleteLocalActivityData), shared on_marker_recorded)
       --> MarkerCommandRecorded;
+    // If we are told to cancel while waiting for the marker, we still need to wait for the marker.
+    WaitingMarkerEvent --(Cancel, on_cancel_requested) --> WaitingMarkerEventCancelled;
+    // Unless, of course, we abandoned the activity, in which case we won't bother with the marker.
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/local_activity_state_machine.rs]: Fixed
